# Diamonds_Project_Supervised_Machine_Learning_with_Scikit-learn

In [21]:
#https://www.kaggle.com/datasets/shivam2503/diamonds

In [22]:
import pandas as pd

df = pd.read_csv("diamonds.csv", index_col=0)
df.head()

carat      cut color clarity  depth  table  price     x     y     z
1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [23]:
df['cut'].unique()

array(['Ideal', 'Premium', 'Good', 'Very Good', 'Fair'], dtype=object)

In [24]:
#transfer from categorical into numerical category
#df['cut'].astype('category').cat.codes()

In [25]:
cut_class_dict = {"Fair": 1, "Good": 2, "Very Good": 3, "Premium": 4, "Ideal": 5}

In [26]:
df['clarity'].unique()

array(['SI2', 'SI1', 'VS1', 'VS2', 'VVS2', 'VVS1', 'I1', 'IF'],
      dtype=object)

In [27]:
clarity_dict = {"I3": 1, "I2": 2, "I1": 3, "SI2": 4, "SI1": 5, "VS2": 6, "VS1": 7, "VVS2": 8, "VVS1": 9, "IF": 10, "FL": 11}

In [28]:
df['color'].unique()

array(['E', 'I', 'J', 'H', 'F', 'G', 'D'], dtype=object)

In [29]:
color_dict = {"J": 1,"I": 2,"H": 3,"G": 4,"F": 5,"E": 6,"D": 7}

In [30]:
#maping
df['cut'] = df['cut'].map(cut_class_dict)
df['clarity'] = df['clarity'].map(clarity_dict)
df['color'] = df['color'].map(color_dict)
df.head()

carat  cut  color  clarity  depth  table  price     x     y     z
1   0.23    5      6        4   61.5   55.0    326  3.95  3.98  2.43
2   0.21    4      6        5   59.8   61.0    326  3.89  3.84  2.31
3   0.23    2      6        7   56.9   65.0    327  4.05  4.07  2.31
4   0.29    4      2        6   62.4   58.0    334  4.20  4.23  2.63
5   0.31    2      1        4   63.3   58.0    335  4.34  4.35  2.75

In [31]:
import sklearn
from sklearn.linear_model import SGDRegressor

df = sklearn.utils.shuffle(df) # always shuffle your data to avoid any biases that may emerge b/c of some order.

X = df.drop("price", axis=1).values
y = df["price"].values

In [32]:
test_size = 200

X_train =http://localhost:8888/notebooks/Machine%20Learning%20with%20Scikit-learn%20-%20Data%20Analysis%20with%20Python%203%20and%20Pandas.ipynb#1.-SGD-Regressor X[:-test_size]
y_train = y[:-test_size]

X_test = X[-test_size:]
y_test = y[-test_size:]

## 1. SGD Regressor

In [33]:
clf = SGDRegressor(max_iter=1000)
clf.fit(X_train, y_train)

print(clf.score(X_test, y_test))

-12201.42133982555


In [34]:
for X,y in list(zip(X_test, y_test))[:10]:
    print(clf.predict([X])[0], y)

-18457.544508688152 9641
352902.5374932587 18777
-74839.38664776087 6770
-572874.6130620316 741
244.67809730023146 4362
-95763.42404150218 542
-23772.878924906254 8145
25503.696477793157 1158
-263658.7035840452 571
-336439.19090004265 1094


### Well, that's not very good. The score for these regression models is r-squared/coefficient of determination, so I am actually not even sure how we got -12201.42133982555, but apparently we did. R-Squared is more often between 0 and 100%, where 100% is a perfect fit (1.0). Let's try support vector regression instead:

## 2. SVR 

In [35]:
from sklearn import svm

clf = svm.SVR()

clf.fit(X_train, y_train)
print(clf.score(X_test, y_test))

-0.1193489836257855


In [36]:
for X,y in list(zip(X_test, y_test))[:10]:
    print(clf.predict([X])[0], y)

2428.9171261639713 9641
2472.925422092153 18777
2514.718699656961 6770
2330.437431692503 741
2361.46199597975 4362
2306.02339824472 542
2403.3520364930573 8145
2312.9959055449945 1158
2342.8536832097957 571
2317.8468381778 1094


## 3. SGD Regressor

In [37]:
clf = SGDRegressor(max_iter=10000)

clf.fit(X_train, y_train)
print(clf.score(X_test, y_test))

for X,y in list(zip(X_test, y_test))[:10]:
    print(clf.predict([X])[0], y)

-88458525.80889536
12461934.348567486 9641
-16078320.633854866 18777
19028630.497744083 6770
57087489.430986404 741
11035509.327397823 4362
22258542.002870083 542
14172368.976502895 8145
10212563.45147705 1158
31593191.80303812 571
36391663.75153828 1094


## 4. SVR 

In [38]:
import sklearn
from sklearn import svm, preprocessing

df = sklearn.utils.shuffle(df) # always shuffle your data to avoid any biases that may emerge b/c of some order.

X = df.drop("price", axis=1).values
X = preprocessing.scale(X)
y = df["price"].values

test_size = 200

X_train = X[:-test_size]
y_train = y[:-test_size]

X_test = X[-test_size:]
y_test = y[-test_size:]

clf = svm.SVR()

clf.fit(X_train, y_train)
print(clf.score(X_test, y_test))

for X,y in list(zip(X_test, y_test))[:10]:
    print(f"model predicts {clf.predict([X])[0]}, real value: {y}")

0.6208589810113894
model predicts 1166.2168994839058, real value: 1046
model predicts 6825.830883351922, real value: 7962
model predicts 5178.068411827111, real value: 5202
model predicts 565.9799479868157, real value: 855
model predicts 5835.809245291588, real value: 6223
model predicts 5653.600968688051, real value: 6694
model predicts 4388.0297803342055, real value: 4374
model predicts 790.8512668391895, real value: 876
model predicts 687.5254343545425, real value: 518
model predicts 6524.4156361549085, real value: 18682


### Preprocessing improved our score a bit, so that's nice. We could keep tweaking things and probably improve this model further. Any new diamond data you got would need to be combined into your main dataset, scaled, then predicted from.